helper/schema/schema.go

```diff
@@ -773,6 +777,16 @@ func (m schemaMap) internalValidate(topSchemaMap schemaMap, attrsOnly bool) erro
 			}
 		}
 
+		if len(v.AllOf) > 0 {
+			if len(v.ExactlyOneOf) > 0 || len(v.AtLeastOneOf) > 0 {
```

Comment:
I don't believe the AtLeastOneOf check should be done because you could use a different set of keys in AtLeastOneOf in addition to the keys you're checking in AllOf. It's pretty wonky but let's take a keys example where you have to put up AtLeastOne key but one of those keys has two parts so you'd throw AtLeastOneOf on all the key related attributes and then throw an extra AllOf block on the two keys that must be specified together. 

The check for ExactlyOneOf is spot on though.

Reply:
From my point of view it makes no sense to combine `AllOf` with any of the other mentioned "operators", because `AllOf` shall guarantee that **all** listed attributes are specified, either explicitly or by a default value. I don't see any reason why this should be combined with `AtLeastOne` or `ExactlyOneOf`. Thus the `if` statement in line 781.

Reply:
To further expand on my example, over in AzureLand we deprecated `connection_string` in favor of `account_name` and `account_key`. I had to do manual checks to get this to work but with this new validation we could instead do

```
  "connection_string": {
    AtLeastOneOf: []string{"connection_string", "account_name", "account_key"},
  },
  "account_name": {
    AtLeastOneOf: []string{"connection_string", "account_name", "account_key"},
    PairedWith: []string{"account_key", "account_name"},
  },
   "account_key": {
    AtLeastOneOf: []string{"connection_string", "account_key", "account_name"},
    PairedWith: []string{"account_name", "account_key"},
  },
```

I'm omitting certain extra validation steps just to illustrates a use case where you must specify one of these three keys but two of them are paired together

Reply:
Thanks for the example. I see the issue here. Because you want to deprecate the `connection_string` you made it optional as well as the other two new properties. So the user might specify none of the above and Terraform wouldn't notice in the `validation` phase. 

But, from my point of view, this type of check you shouldn't do on properties because this would mix the semantics between a validation for a specific property and the validation for a complete schema of a resource (data source). 

My understanding (until now) for all those "operators" like `AtLeastOneOf` or `ConflictsWith` is that they are scoped to the specific property where they are defined and with the line `AtLeastOneOf: []string{"connection_string", "account_key", "account_name"},` at the property `account_key` you would violate this semantic, because you might require the `connection_string` to be set although it (strictly saying) **collides** with the `account_key` attribute.

Reply:
The nice thing about all these validations is that we can daisy chain them together. I specifically omitted ConflictsWith but we could add that here for sure to get a more complete picture. I think the thing to note is that we would have to write this logic no matter what to get the desired effect and it's be best to get that logic written during plan -time rather than erroring out mid Terraform run.

```
  "connection_string": {
    AtLeastOneOf: []string{"connection_string", "account_name", "account_key"},
    ConflictsWith: []string{"account_name", "account_key"},
  },
  "account_name": {
    AtLeastOneOf: []string{"connection_string", "account_name", "account_key"},
    AllOf: []string{"account_key", "account_name"},
    ConflictsWith: []string{"connection_string"},
  },
   "account_key": {
    AtLeastOneOf: []string{"connection_string", "account_key", "account_name"},
    AllOf: []string{"account_name", "account_key"},
    ConflictsWith: []string{"connection_string"},
  },
```

Reply:
Do you mind further elaborating on how we would violate the semantic? Examples would definitely help me see your point of view

Reply:
Also, this is the code I had to write that the above would solve in a much cleaner way. https://github.com/terraform-providers/terraform-provider-azurerm/pull/6304/files#diff-23333db7a329c446abb71d2877f1bcf1R733

Reply:
> 
> 
> Do you mind further elaborating on how we would violate the semantic? Examples would definitely help me see your point of view

Well from my point of view we're talking about two different things here:

1. The (validation) requirements for a specific property and
2. The (validation) requirements for the schema as a whole

Your proposal combining `AtLeastOneOf` with `AllOf` to ensure that at least one attribute of `connection_string` and `account_name` violates the semantic in a way that you try to validate the schema on an attribute validation. I know this might be pedantic but I don't like it when concerns are mixed together. 

So ideally we should have some sort of validation function for the schema: typically are all properties are set, especially considering all possible alternatives (`connection_string` or `account_name`, `account_key`. (It's something like with the possible (valid) parameter combinations when you use ParameterSets in Powershell.) And the validations functions for the properties (which are already in place). 

I think this provides a concise way of how to use which validation function on what concern (schema, property), instead of wildly combine validations functions on the properties where the user not only must understand the semantic of `AtLeastOneOf` etc., but also the semantics how they interact together to ensure a valid schema. Not talking about the poor person documenting this.

Reply:
One additional word: because introducing a schema validation would impose a major change of the SDK, I could definitely live with the solution proposed by you above. But I would suggest that someone else from the SDK Team at Hashicorp (perhaps @paultyng) would give his or hers opinion about this.

Reply:
Hello @tmeckel thank you for the contribution! The SDK as its written today unfortunately doesn't naturally support whole resource validation (unless dealing with CustomizeDiff.... let's not go there), these special behaviors are our current best effort way of creating groups or sets of keys that enable plan time validation that would otherwise be done manually and during apply time.

@mbfrahry is right that these behaviors are meant to all work together. For example it would be great to be able to enforce `ExactlyOneOf` and then for a particular attribute of that grouping also enforce `AllOf`. I would like to suggest another name though, perhaps `RequiredWith`, that might align closer with the goal here.

If you were looking to enforce attributes being set across a whole resource, you would just mark them as required. As I understand it, this will give us the ability to enforce requirement only when one of the grouping is set (then the rest must also be set), otherwise they are marked `Optional: true`.

Reply:
@appilon Thanks for expressing you opinion here and giving some more explanation on how the various attributes are used or shall be used together.

As I wrote before, I am aware that introducing a way to validating the schema as a whole would mean a massive change of the SDK, and also from the fact that I don't slavishly cling to the if statement, I would suggest that I rename the property `AllOf` to `RequiredWith` and remove the if statement. 

If you guys fine with this, I can provide a changed PR today.

Reply:
👍 thanks @tmeckel this would be ideal, once those changes are made we will prioritize getting this PR merged.

Reply:
@appilon @mbfrahry As planned I push the changes to the PR.